### ISSUE TITLE
CalDAV Task support

### ISSUE THREAD

**Issue by Akselmo** [FossifyOrg/Calendar#36]
**Checklist**
 - [x] I made sure that there are no existing issues - open or closed - to which I could contribute my information.
 - [x]  I have read the FAQ and my problem isn't listed.
 - [x]  I have taken the time to fill in all the required details. I understand that the feature request will be dismissed otherwise.
 - [x]  This issue contains only one feature request.
 - [x]  I have read and understood the [contribution guidelines.](https://github.com/FossifyOrg/General-Discussion?tab=readme-ov-file#faq)

**Is your feature request related to a problem? Please describe.**
I would like to be able to modify my CalDAV tasks from this app. I know there's a local only tasklist in the app, but I use tasks across devices. For example Nextcloud Tasks. 

**Describe the solution you'd like**
When I create a task in the app, I want it to sync with the other CalDAV calendars I have. 

**Describe alternatives you've considered**
I have a separate tasks app but usually those are full of features I don't need. Having both in one calendar app would be way more suitable for me.

**Comment by inson1** [comment]
https://github.com/SimpleMobileTools/Simple-Calendar/issues/1742

**Comment by Akselmo** [comment]
To give more context, basically the application would have to handle `VTODO` components: https://icalendar.org/iCalendar-RFC-5545/3-6-2-to-do-component.html

**Comment by antrmn** [comment]
To make things simpler, it could simply read tasks from some external app which provides its own API. Tasks.org provides its own content provider API which allows to access tasks from an external app on the same device.

**Comment by Daybayzayd** [comment]
I was so disapointed when all the tasks i created in thunderbird were not synchronized in my calendar app.. :'(

**Comment by Daybayzayd** [comment]
> [SimpleMobileTools/Simple-Calendar#1742](https://github.com/SimpleMobileTools/Simple-Calendar/issues/1742)

There is no caldav task support in their calendar.

**Comment by KettLovahr** [comment]
Yes please! I recently started organising things locally, hosting my own CalDAV server at home, and it would be wonderful to have tasks synchronised along with events.

**Comment by achadwick** [comment]
Yes, please! I might replace [/e/OS's fork](https://doc.e.foundation/what-s-e#eos-default-applications) of [OpenTasks](https://f-droid.org/packages/org.dmfs.tasks/) with something integrated if it was possible (currently I'm just playing with Fossify Calendar). I'd be happy to test a tasks feature out against my NextCloud accounts. I think their OpenTasks fork uses some flavour of [EteSync](https://f-droid.org/en/packages/com.etesync.syncadapter/) behind the scenes, buried behind the bundled accounts manager.

A nice feature of the /e/OS Tasks app (at least) is that it supports markdown in the description - or at least, the checklist and bullet list syntax. It derives the completion %age from the checked boxes. It'd be a really nice feature to have, but that's a big wish.

**Comment by wvffle** [comment]
Is there any ongoing or planned work regarding this feature? It seems like a lot of people are missing it, myself included.

**Comment by naveensingh** [comment]
No progress yet. I do plan to work on it. We'll probably have to contact DAVx5 devs to include FC as a supported tasks app.

**Comment by kaafir786** [comment]
Any progress by now?

**Comment by bemyak** [comment]
FWIW, there's an incomplete implementation of VTASK support in Etar here: https://github.com/Etar-Group/Etar-Calendar/pull/1279

Hope it helps :pray:

**Comment by vtrincadoh** [comment]
I could really use this feature. Is there any work presently that needs to be done?

**Comment by kaafir786** [comment]
> I could really use this feature. Is there any pending work presently that needs to be done?

You can manage task locally but caldav tasks (Calendar synced with davx⁵,decsync etc) are not supported yet. You cannot view, add, manage them.

**Comment by Baloe2000** [comment]
Is there already an update on the implimentation of this feauture? There are currently no open source calendars with this and I think it would be great to see this in fossify calendar! 😄

**Comment by abishekmuthian** [comment]
I'm using jtx Board with nextcloud CalDAV and it works flawlessly with other CalDAV connected clients. I wish @naveensingh the best to implement the feature when ever possible, Until then jtx board works well for me.

**Comment by Konkrad** [comment]
Any update on this? Would be willing to chip in a few dollars into a bounty for this issue. Not sure which tools is normally used for this.

**Comment by loa-hong-cong** [comment]
@naveensingh 
Hi, thank you very much for making such a great suite of apps, i've been using them for less than a month but i've loved all of your applications so far!

I would like to vouch for this issue, this would make my life much easier if it was fixed (seriously haha)
I have donated $5 to show support for this issue, I know its not much
![Image](https://github.com/user-attachments/assets/0a469628-4643-427e-955a-269d6d642671)

 Please let me know if I can be of any help, I know some Dart & Flutter, which is somewhat similar to Kotlin

**Comment by egwfwegw** [comment]
Hi, I am @loa-hong-cong (my account got banned) for some reason

Im trying to grasp the codebase so I can potentially make a PR for this issue

Would I be correct in saying that I would have to change `org.fossify.calendar.receivers.CalDAVSyncReceiver` to support the VTODO spec?

**Comment by naveensingh** [comment]
Yes but it involves more than that.

**Comment by DarioBiscevic** [comment]
Shouldn't it be relatively easy to implement? Fossify Calendar already knows how to handle VTODOs from .ics files, the only thing that seems to be missing is the "inclusion" of tasks among the objects to actually sync with the remote.
This is really the only feature that's missing to make Fossify Calendar a complete replacement for Google Calendar

**Comment by notjiatan** [comment]
Bump, I really hope this issue gets fixed, it would be invaluable to me

**Comment by handkerchief333** [comment]
The same applies to me. This feature is really wanted and needed by many people. Thank you very much for your efforts.

**Comment by momonator25** [comment]
I started using Fossify Calendar Today which works nice with DAVx⁵ and my self hosted Nextcloud and would be more then happy to see this feature in the future! :)

**Comment by liampas** [comment]
I also think this is a must feature

**Comment by notjiatan** [comment]
My life would be some much easier with this feature!

**Comment by christian179717** [comment]
I don't think it's very helpful if another 20 guys post how helpful the feature might be. Everyone gets notified, distracted from work and hopes there is progress in development, but there's nothing.
If you want to express that you want the feature, you can react to the issue with thumbs up or something else.

**Comment by bcspragu** [comment]
Echoing [my comment](https://github.com/SimpleMobileTools/Simple-Calendar/issues/1742#issuecomment-3189458730) from the SimpleMobileTools repo -- if nobody is working on this, I'd be willing to give it a shot with a few pointers. I've recently set up my own self-hosted CalDAV (Radicale) instance and this is the last missing piece for me.

**Comment by duub** [comment]
Hi! good to know I'm not the only one stuck with this issue.

**Comment by ororktn** [comment]
Disclaimer: i'm not an Android developer, i'm just interested syncing my tasks on FossifyCalendar.

Ok so I started to think about how to do that. From what i understood, it can be a bit more tedious than it seems:
- Fossify Calendar uses Android API to sync Calendars through `CalendarContract`. For ICS VTODO imports, they are parsed manually from the file.
- DAVx5 compatible task apps more or less all implement their own API (Contract) that DAVx5 uses to sync tasks.

Would it be possible to find a way to use DAVx5 to transfer the data through Android `CalendarContract`? I don't think so, but someone may have an idea on how to do it. Otherwise, i feel like we're doomed to use one of the other apps implementation (kinda risky and could change any time), or make our own `Contract`, that would then need to be used by DAVx5 or other synchronization tools.

**Comment by rienheuver** [comment]
@bcspragu seeing as @naveensingh just marked this featue request as accepted, perhaps it's time to give it a shot if you still have the spare time? I still believe a great calendar (like Fossify) with propering davx5 integration including recurring tasks would attract quite a user base since loads of people are trying to escape Google right now, but still would love to have (recurring) tasks and calendar properly integrated

**Comment by iosonofabio** [comment]
Thanks for the awesome app. I'd like to look into this as well and see how hard it would be to make a PR.

**Comment by PJFonseca** [comment]
> [@bcspragu](https://github.com/bcspragu) seeing as [@naveensingh](https://github.com/naveensingh) just marked this featue request as accepted, perhaps it's time to give it a shot if you still have the spare time? I still believe a great calendar (like Fossify) with propering davx5 integration including recurring tasks would attract quite a user base since loads of people are trying to escape Google right now, but still would love to have (recurring) tasks and calendar properly integrated

I'm one of those. Self-hosting on a Synology, I struggled to find a proper app that syncs events + birthdays + tasks, until today. Finally!

I can sponsor a developer to bring this feature to this app and benefit the entire FOSS community. @naveensingh you lead the way and tell me who should I sponsor :)

**Comment by iosonofabio** [comment]
I'm not an Android dev by day, so if @bcspragu or @naveensingh you happen to have time for this, please take this! Happy to help in my free time however I can.

If it helps at all, I took a look at how DAVx5 does this and it's quite special. They use a few self-developed, lower-level dependencies to sync things which you can read on their [README](https://github.com/bitfireAT/davx5-ose), however task parsing requires [one more import](https://github.com/bitfireAT/davx5-ose/blob/24d4ba65e512e7cd73daf48da521aab1040f455c/app/src/main/kotlin/at/bitfire/davdroid/sync/TasksSyncManager.kt#L28) from their older, now archived [ical4android](https://github.com/bitfireAT/ical4android) library:

https://github.com/bitfireAT/ical4android/blob/01bce4a8ee8395402a53c06263839ce3afe97e2b/lib/src/main/kotlin/at/bitfire/ical4android/Task.kt#L113

Though I could not dig vertically to the bottom, the very next `write` function uses [ical4j](https://github.com/ical4j/ical4j) as a dependency to actually construct valid iCal entries. The rabbit hole ends [here](https://github.com/ical4j/ical4j/blob/5ac389f4c0de7bd2af4503db28c1d33705201bb7/src/main/java/net/fortuna/ical4j/model/Calendar.java#L227).

The impression is that even on the DAVx5 side of things syncing tasks is at the trailing end of updates, same as here. I'll see if I can understand better how Fossify Calendar does the parsing/writing of VTODOs.

**Comment by bcspragu** [comment]
Thanks for the pointers @iosonofabio ! I'll give this a shot next weekend. I'm also not an Android dev by day, though I do (server-side) Kotlin stuff for work now, which has gotta count for something. I'll report back with what I find

**Comment by freeDom-** [comment]
~~AFAIK DAVX5 is not syncing tasks at all, which is why a third party app like Tasks.org, OpenTasks or jtxboard is required to do so.~~

OpenTasks is discontinued and Tasks.org now requires a subscription for syncing (which is ok for me, but not ideal, because its not really foss).
I do not know for sure, but I think that jtxboard does not provide an API for syncing with other apps such as Tasks.org.

So, this is probably why this topic is harder than it sounds... One would need to implement the syncing feature from bottom to top in a calendar app, which currently utilizes DAVX5 for syncing.
Ideally, this feature should rather be implemented on DAVX5 so that any calendar app could benefit from it, right?

TLDR: It seems that tasks must be handled differently for some reason than events. **So what DAVX5 does is syncing tasks to the database of a supported app.** Therefore, it definitely would be easiest to become a supported app of DAVX5.

**Comment by iosonofabio** [comment]
On my phone DAVX syncs tasks via caldav, I'm not sure whether all features (e.g. recurrent tasks) are handled correctly but it does seem to do that, so one could rely on that perhaps.

**Comment by tropicalsriracha** [comment]
DAVX5 sync tasks with caldav

You can use DAVX5 with Tasks.org as middleman or caldav with tasks directly.

[From their website](https://tasks.org/docs/davx5.html)

DAVx⁵ vs CalDAV

Both options provide nearly identical functionality, with the exception that Tasks cannot create, rename, colorize, or delete your CalDAV lists when synchronizing with DAVx⁵.

**Comment by karan** [comment]
Yeah I've been using davx5 with tasks.org and it works fine.

**Comment by 7thgenerationdesign** [comment]
Dominik, I believe you are incorrect on multiple points:

- DAVx5 is required to sync tasks between a CalDAV server and a client like Tasks.org, not the other way around
- Tasks.org does not require a subscription to sync, they only require a subscription if you want to use THEIR CalDAV server. Using your own, or a third party server like Google, The Good Cloud, or something else is completely free

Mar 30, 2026 11:21:35 AM Dominik ***@***.***>:

**Comment by iwan-uschka** [comment]
Can confirm @freeDom-'s statement. Just installed https://github.com/tasks/tasks via Google Play Store and had to make a $ubscription to be able to use DAVX5.
Thanks for mentioning https://github.com/tasks/tasks. I am not a fan of having 2 separate apps handling my calendar stuff (events and tasks/reminders) but it's cool to have full access now!

**Comment by MusterGit** [comment]
> TLDR: It seems that tasks must be handled differently for some reason than events. **So what DAVX5 does is syncing tasks to the database of a supported app.** Therefore, it definitely would be easiest to become a supported app of DAVX5.

Here is a nice illustration from their website:

[![](https://www.davx5.com/fileadmin/user_upload/faq/how_davx5_works_old.png)](https://www.davx5.com/fileadmin/user_upload/faq/how_davx5_works_old.png)

Can someone explain to me why tasks cannot be synced to Android's Calendar Storage as well if CalDAV is used for both calender events and tasks?

**Comment by 7thgenerationdesign** [comment]
Good find on that illustration!

FWIW, [acalendar](https://acalendar.tapirapps.de/en/support/solutions/articles/36000058826-tasks-feature) offers the functionality desired. The docs infer, if I understand them correctly, that the functionality is achieved by syncing with the tasks apps themselves (which are synced using DAVx5 or whatever proprietary sync method).

**Comment by freeDom-** [comment]
> > TLDR: It seems that tasks must be handled differently for some reason than events. **So what DAVX5 does is syncing tasks to the database of a supported app.** Therefore, it definitely would be easiest to become a supported app of DAVX5.
> 
> Here is a nice illustration from their website:
> 
> [![](https://www.davx5.com/fileadmin/user_upload/faq/how_davx5_works_old.png)](https://www.davx5.com/fileadmin/user_upload/faq/how_davx5_works_old.png)
> 
> Can someone explain to me why tasks cannot be synced to Android's Calendar Storage as well if CalDAV is used for both calender events and tasks?

Android does not have a standardized, system-wide tasks provider like it does for contacts or calendars. Unlike calendars, Android itself doesn’t offer a universal, built-in tasks database or API that all apps can read from or write to.

This is why you need an app that can handle the tasks and provides a place (database) to store them. And this is also why it is not trivial to implement them in a calendar app.

acalendar is utilizing Tasks.org, but I do not know if Tasks.org is providing an API to access the locally stored tasks or if it provides access to the tasks on the caldav server itself.

**Comment by abaker** [comment]
Android's Calendar Storage model is influenced by iCalendar, but doesn't really have anything to do with CalDAV, hence the need for third-party sync adapters like DAVx5. The Google Calendar app uses proprietary sync APIs, not CalDAV, when synchronizing your Google Calendars to your phone.

> acalendar is utilizing Tasks.org, but I do not know if Tasks.org is providing an API to access the locally stored tasks or if it provides access to the tasks on the caldav server itself.

Tasks.org embeds an OpenTasks content provider copied straight out of the OpenTasks app. DAVx5 writes to this content provider during sync, and other apps, like aCalendar, are free to read from this content provider as well. This content provider only exposes data from DAVx5, as the other sync options don't use the OpenTasks database.

I was asked to weigh in on this thread, I promise I didn't come to plug tasks.org 😎 But if Fossify Calendar wanted to read from the OpenTasks content provider, or wanted a custom content provider to access all tasks.org's data, I would be happy to help!

> Tasks.org only offers free synchronisation for Google and Microsoft calendars any longer.

Just to clarify, Tasks.org has always required a $0.99 USD/year subscription for CalDAV and DAVx5 sync when installed from Google Play; it was never free. This is a common misconception when people inadvertently switch from F-Droid to Google Play and think I did a rug pull 🥲

**Comment by MusterGit** [comment]
> Android does not have a standardized, system-wide tasks provider like it does for contacts or calendars. Unlike calendars, Android itself doesn’t offer a universal, built-in tasks database or API that all apps can read from or write to.

Sorry for being possibly naive but can the Calendar Storage be repurposed for tasks so that there is at least a robust interface for any 3rd party task app? This may sound hacky but not as hacky as injecting entries in foreign databases.

**Comment by freeDom-** [comment]
> > Android does not have a standardized, system-wide tasks provider like it does for contacts or calendars. Unlike calendars, Android itself doesn’t offer a universal, built-in tasks database or API that all apps can read from or write to.
> 
> Sorry for being possibly naive but can the Calendar Storage be repurposed for tasks so that there is at least a robust interface for any 3rd party task app? This may sound hacky but not as hacky as injecting entries in foreign databases.

Not a good design. If something is going to be implemented, it should not be something hacky at all. Also, the calendar storage is most probably a database as well.

As @abaker explained and seen in the diagram above, there is a content provider to handle task reads properly.
How about storing data? From the diagram it looks as if the OpenTasks task provider can be used in both directions and Davx5 is doing the caldav synchronization? How about the Tasks.org based tasks provider? Can this only be read? Is acalendar then using the caldav API directly to store the tasks on the server?

It sounds like the OpenTasks content provider is the way to go. Either by implementing a dedicated one for fossify calendar, or by reusing an existing one (e.g. from tasks.org). I am going to look at a reference implementation to see if it takes a lot of work to get a PR ready.

**Comment by MusterGit** [comment]
> It sounds like the OpenTasks content provider is the way to go.

OpenTasks has not been touched for 5 years. Seems like a robust enough interface 😉 
Seriously, good idea! 🚀

**Comment by freeDom-** [comment]
> > It sounds like the OpenTasks content provider is the way to go.
> 
> OpenTasks has not been touched for 5 years. Seems like a robust enough interface 😉 Seriously, good idea! 🚀

I did not mean to use the one from OpenTasks, because the app is not even available on Googly Store for recent android versions.

> Either by implementing a dedicated one for fossify calendar, or by reusing an existing one (e.g. from tasks.org).

**Comment by thomas2403** [comment]
There are some forks of opentask, this seems to be still in development, last release was one month ago. It is working for me with A15. It is not available on Play Store or F-Droid though.

https://github.com/nosp7m/tasksandlists

**Comment by rapidmountain** [comment]
There is an app called "Business Calendar 2" for Android that supports both Calendars and tasks. It shows the tasks on the calendars too. I use it together with Nextcloud and a O365 account. 

But the Fossify Calendar has a sleeker UI and I would love to get tasks added. Perhaps you can look at how Business Calendar 2 has implemented it? They have a native CalDav integration as part of their Pro version (paid).

**Comment by freeDom-** [comment]
> Perhaps you can look at how Business Calendar 2 has implemented it? They have a native CalDav integration as part of their Pro version (paid).

It is closed source - we do not have access to the source code.

**Comment by eshattow** [comment]
Adding a hint for people looking for a comparable app with support of the VTODO and VJOURNAL standards, have a look at jtx Board https://f-droid.org/packages/at.techbee.jtx/ although it is not an answer for the message of this issue, it is that more full-featured task and journal oriented app not the lightweight todo enhancement being requested.

**Comment by imagoiq** [comment]
Another hint for users of "Tasks". I discovered that the launcher [Kvaesitso](https://kvaesitso.mm20.de/) can combine tasks and events in the calendar widget.

**Comment by 7thgenerationdesign** [comment]
Fantastic find - thank you for sharing!!

Apr 28, 2026 10:36:26 PM Loïc Fürhoff ***@***.***>:

**Comment by arreme** [comment]
I would also love to have this feature if possible! There is a similar app called aCalendar which I think it implements it very well!

**Comment by hashblen** [comment]
I have done some research on this, and plan on attempting at implementing it soon(TM). Fair warning, I'm not an Android dev at all, so if I get something working this will probably need thorough review from people who actually know what they're doing here.

Android doesn't have a CalendarContract-equivalent for tasks, so there's nothing built-in for us to sync against like we do for events. The workaround other apps use (aCalendar+, for example) is to connect to a separate app's task provider instead (tasks.org in their case) and let DAVx5 sync that. So the user ends up with three apps: Calendar, a task provider app, and DAVx5.
![DAVx5 Interaction Architecture](https://manual.davx5.com/_images/how_davx5_interacts_with_other_components.png)
So Fossify never needs to implement CalDAV/auth, it only needs to read/write jtx Board's provider (and later I might add task.org's provider) the same way it already reads/writes CalendarContract for events.

The provider contract is documented via jtx Board's own instrumented tests: [SyncContentProviderTest.kt](https://github.com/TechbeeAT/jtxBoard/blob/develop/app/src/androidTest/java/at/techbee/jtx/SyncContentProviderTest.kt), and the client-facing API surface is [JtxContract](https://github.com/TechbeeAT/jtxBoard/blob/develop/app/src/main/java/at/techbee/jtx/contract/JtxContract.kt) ([dokka](https://bitfireat.github.io/ical4android/ical4android/at.techbee.jtx/-jtx-contract/index.html))
Once this lands, the same shape can support [tasks.org](https://tasks.org/)as a second backend. I'm thinking a small ExternalTaskProvider interface or similar, so jtx Board and tasks.org are two implementations rather than two parallel codepaths.

For a first pass I'm thinking:
* detect whether jtx Board is installed
* add a "connect task provider" setting
* new source prefix "Jtx-<collectionId>", following the existing "Caldav-<id>" convention on Event.source
* read + two-way sync for VTODO fields that map cleanly onto the current Event model: title, description, due date, completed state

jtx Board also has priority, percent-complete, and related-to (subtasks). Those don't have a home in Event yet, so I'd leave them for a follow-up if we want to add them. For jtx-specific fields we won't have UI for (categories, resources, attachments, subtasks), the plan is to preserve them on round-trip without exposing them, plus add a shortcut to open the task directly in jtx Board when someone needs full editing.

I am open to any ideas and discussion about this, do not hesitate.
